### DISCUSSION TITLE
how to get pod or node metrics info

### DISCUSSION THREAD

**Discussion by vCassius** [kube-rs/kube#1315] [Q&A] [answered]
anyway to get pod or node metrics info? when i show pod details i need to select pod's metrics info and drew it use vue frontend.

**Comment by mateiidavid** [comment]
It depends on what you need the metrics for, probably. If your cluster supports it, you could try using the [`metrics-server api`](https://github.com/kubernetes-sigs/metrics-server). I haven't looked into it too much, but it seems like [k8s-open-api](https://arnavion.github.io/k8s-openapi/v0.20.x/k8s_openapi/api/autoscaling/v2/index.html) has some autoscaling types that seem to coincide with the metrics-server api. Note the metrics-server api usecases though, it might not be a good fit for what you need. `kube-rs` can be used to build a client that gets the autoscaling metrics directly from the k8s API server

An alternative: use prometheus to ingest metrics and use a prometheus client. Work directly on the timeseries data to get a more accurate view.

Hope that helps!

**Reply by vCassius** [reply]
thx guy

**Reply by mateiidavid** [reply, marked as answer]
Here's a different way to do it with a DynamicClient, this would be the equivalent of `kubectl top`

```rust
use kube::{core::DynamicObject, Api, Resource, ResourceExt};
use serde::Deserialize;

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let client = kube::Client::try_default().await?;
    let apigroup = kube::discovery::group(&client, "metrics.k8s.io").await?;
    let res = apigroup.recommended_resources();
    for (ar, _caps) in res {
        println!("Ar: {ar:?}");
    }
    let (ar, _caps) = apigroup.recommended_kind("NodeMetrics").unwrap();
    let api: Api<DynamicObject> = Api::all_with(client.clone(), &ar);
    for obj in api.list(&Default::default()).await? {
        let (name, metric) = extract_metric(obj)?;
        println!("{name}: {metric:?}");
    }
    Ok(())
}

#[derive(Default, Deserialize, Clone, Debug)]
struct Metric {
    cpu: String,
    memory: String,
}

fn extract_metric(obj: DynamicObject) -> anyhow::Result<(String, Metric)> {
    let metric = obj
        .data
        .get("usage")
        .cloned()
        .ok_or_else(|| anyhow::anyhow!("missing 'usage'"))?;
    let metric = serde_json::from_value(metric)?;
    Ok((obj.name_any(), metric))
}


```

**Reply by vCassius** [reply]
thx guy. It's very useful to me!
